### ISSUE TITLE
Controversial: `never_loop` lint should be allowed

### ISSUE THREAD

**Issue by ElliottjPierce** [rust-lang/rust-clippy#17511]
### Description

The [`never_loop`](https://rust-lang.github.io/rust-clippy/master/index.html?search=never_l#never_loop) lint should be allowed by default (or at least not denied). Why? I've found that a loop that never loops is very useful and is in fact even *necessary* for some things.

**Usefulness**

Consider this example:

```rust
let result = if a {
     // Some large work
} else {
    let c = // Some large work
    let b = // Some large work
    if b {
        // Some large work that uses `c`
    } else {
        // Some large work that uses `c`
    }
}
```

There isn't an easy way to get rid of that pesky nesting. You could argue that at a certain point, this should be broken into helper functions, but that's not always feasible.

But a never loop fixes this:

```rust
let result = loop {
    if a {
        break // Some large work
    }

    let c = // Some large work
    let b = // Some large work
    if b {
        break // Some large work that uses `c`
    }

    break // Some large work that uses `c`
}
```

That's much nicer! The loop is a little annoying because you have to make extra sure that all code paths break. Maybe a `do` syntax could make this better in the future:

```rust
let result = do {
    if a {
        break // Some large work
    }

    let c = // Some large work
    let b = // Some large work
    if b {
        break // Some large work that uses `c`
    }

   // Some large work that uses `c`
}
```

But for right now, the never loop works where you need it. I wouldn't argue that this is universally preferable to an extra layer of nesting, but I would argue that there are times when deep enough nesting and tricky enough function extraction make this a defensible option. (Sometimes it's hard to extract a helper function if it needs very many parameters, deals with unsafe invariants that become less obvious when not inline, makes statefulness harder to follow when not inline, etc.)

**Necessary**

I decided to write this after finding a case where a "never loop" was the only reasonable solution. 
(For background, I'm trying to learn wgpu, so being a beginner here, there's probably a more elegant solution to this, but the syntax is still interesting.)

I'm trying to initialize an `Arc<Mutex<Option<T>>>` but that `T` (in this case a render surface) has to be made on the main thread, the pipework for which I've implemented via async. The whole thing is in an async function that must be `Send`, so I can't hold the `MutexGuard` across an await point. This is non-trivial to pull into a helper function. (Even if it's not, pretend with me for sake of argument.) Here's the relevant snippet:

```rust
let surface = {
    let mut render_surface = render
        .surface
        .lock()
        .unwrap_or_else(PoisonError::into_inner);

    // (hidden): Try to reuse existing surface here.

    if let Some(surface) = render_surface.as_ref() {
        surface.clone()
    } else {
        let window_clone = render.window.clone();
        let instance_clone = self.instance.clone();

        // IMPORTANT: We must drop this mutex guard before awaiting so that the future is `Send`.
        drop(render_surface);

        let render_surface = render.surface.clone();
        let Some((surface, _window)) = app
            .run_on_main(move |_app, _| {
                let Ok(handle) = window_clone.window_handle() else {
                    return None;
                };
    
                let surface = {
                    // (hidden) Make surface here
                };
    
                let mut render_surface = render_surface
                    .lock()
                    .unwrap_or_else(PoisonError::into_inner);
                let surface = Arc::new(surface);
                *render_surface = Some(surface.clone());
                Some((surface, window_clone))
            })
            .await
            .ok()
            .flatten()
        else {
            return Err(ErrCantFindGPU::Context);
        };
        surface
    }
};
```

Only problem was, this didn't work. Despite the mutex guard being dropped, the compiler still won't implement `Send` on the future. Maybe this is a problem with the compiler being a bit too careful, but something tells me there's a good reason for it. Everything I tried had the same problem, until I put it in a loop.
 
```rust
let surface = loop {
    {
        let mut render_surface = render
            .surface
            .lock()
            .unwrap_or_else(PoisonError::into_inner);

        // (hidden): Try to reuse existing surface here.

        if let Some(surface) = render_surface.as_ref() {
            break surface.clone();
        }
    }

    let window_clone = render.window.clone();
    let instance_clone = self.instance.clone();
    let render_surface = render.surface.clone();
    let Some((surface, _window)) = app
        .run_on_main(move |_app, _| {
            let Ok(handle) = window_clone.window_handle() else {
                return None;
            };

            let surface = {
                // (hidden) Make surface here
            };

            let mut render_surface = render_surface
                .lock()
                .unwrap_or_else(PoisonError::into_inner);
            let surface = Arc::new(surface);
            *render_surface = Some(surface.clone());
            Some((surface, window_clone))
        })
        .await
        .ok()
        .flatten()
    else {
        return Err(ErrCantFindGPU::Context);
    };
    break surface;
};
```

And then the error went away. That's why I say the never loop was necessary. (Though, maybe there's something else I could have done??)

Anyway, I ended up changing the lint from "deny" to "allow". Ordinarily, I'd leave it there, but I'm opening this issue because, at least in my case, it seems the "never loop" was necessary, and IMO, a necessary and simple solution shouldn't be "deny" be default, even if it only solves somewhat niche problems. 

This lint is probably useful to stop beginner mistakes. Maybe it should be "warn" by default? I don't know, but I wanted to bring it to y'all's attention. Maybe it's just me.

### Version

```text
rust 1.97 (and since 1.29)
```

### Additional Labels

@rustbot label +C-enhancement

**Comment by Gri-ffin** [comment]
For the usefulness part, there are many ways to refractor that code, but let's say that we really want that shape, why not use labeled blocks?
```rust
let result = 'result: {
    if a {
        break 'result // Some large work
    }

    let c = // Some large work
    let b = // Some large work
    if b {
        break 'result // Some large work that uses `c`
    }

    // Some large work that uses `c`
}
```

For the second part, the fix from what I can see is not the loop per se, it's that you scoped the guard with an inner block, a better pattern would be to scope the mutex guard using an `Option/match` pattern (or you can just simply early return if there would be no other use for surface):
```rust
let some_surface = {
    let render_surface = render
    .surface
    .lock()
    .unwrap_or_else(PoisonError::into_inner);

    // (hidden): Try to reuse existing surface here.

    render_surface.as_ref().cloned()
};

let surface = match some_surface {
    Some(surface) => surface,
    None => {
        // your logic...
    }
};
```
using a loop that actually never loops just makes the intention of the code confusing.

**Comment by Jarcho** [comment]
Closing since this just uses loops as labelled blocks.

Your loop code compiles because you restructured to remove the explicit drop. The complier's lowering of generators isn't really great when it comes to deciding what is live across await points. The only thing which reliably stops something from being held is for it's scope to end.

**Comment by ElliottjPierce** [comment]
Thanks guys! Yeah, the only way I tried that worked was an extra scope, but you can't put the `if` and the `else` in different scopes.

The solution is in fact labels. How am I just now discovering that labels can be used on any block, not just loops? That's awesome! (I love Rust.)

It might be worth adding this trick to the lint to teach people. The *only* reason you'd want a never loop is to make use of `break`, but if you can do that without the loop (thanks to labels), maybe clippy can suggest a label instead. (IDK if that's actually worth the effort to add, but it would be nice if clippy could teach this part of Rust too.)
